File: backend_kapda_junction/src/models/Order.js
const mongoose = require('mongoose');

const orderItemSchema = new mongoose.Schema({
  product: { type: mongoose.Schema.Types.ObjectId, ref: 'Product' },
  name: String,
  price: Number,
  quantity: Number,
  size: String,
  color: String
});

const orderSchema = new mongoose.Schema({
  user: { type: mongoose.Schema.Types.ObjectId, ref: 'User' },
  items: [orderItemSchema],
  status: { type: String, enum: ['pending', 'confirmed', 'shipped', 'delivered', 'cancelled'], default: 'pending' },
  subtotal: { type: Number, default: null },
  discountAmount: { type: Number, default: 0 },
  coupon: { type: mongoose.Schema.Types.ObjectId, ref: 'Coupon', default: null },
  couponCode: { type: String, default: '' },
  totalAmount: { type: Number, required: true },
  shippingAddress: { type: mongoose.Schema.Types.Mixed, required: true },
  paymentMethod: { type: String, default: 'razorpay' },
  paymentStatus: { type: String, enum: ['pending', 'paid', 'failed', 'refunded', 'partially_refunded'], default: 'pending' },
  /** Set after reduceStock on successful capture — used so refund.processed does not inflate stock if nothing was deducted */
  stockDeducted: { type: Boolean, default: false },
  razorpayOrderId: String,
  razorpayPaymentId: String,
  razorpaySignature: String,
  // Refund tracking
  razorpayRefundId: String,
  refundStatus: { type: String, enum: ['', 'pending', 'processed', 'failed'], default: '' },
  refundAmount: Number,
  /** Last Razorpay / refund error text (audit) */
  refundLastError: { type: String, default: '' },
  /** Admin-only note when refund is manual or retried */
  refundNote: { type: String, default: '' },
  cancelReason: String,
  cancelledAt: Date,
  cancelledBy: { type: mongoose.Schema.Types.ObjectId, ref: 'User' },
  /** Manual post / India Post — no courier API */
  shippingCarrier: { type: String, default: 'India Post' },
  /** Article / consignment / tracking number */
  shippingAwb: { type: String, default: '' },
  /** If India Post URL format changes, admin can paste full tracking URL */
  shippingTrackingUrlOverride: { type: String, default: '' },
  shippedAt: { type: Date, default: null }
}, { timestamps: true });

module.exports = mongoose.model('Order', orderSchema);

File: flutter_kapda_junction/lib/domain/entities/order.dart
import 'package:equatable/equatable.dart';

class OrderItem extends Equatable {
  /// Line id from API (`items[]. _id`) when present — used for returns.
  final String? lineItemId;
  final String productId;
  final String name;
  final double price;
  final int quantity;
  final String? size;
  final String? color;
  final String? image;

  const OrderItem({
    this.lineItemId,
    required this.productId,
    required this.name,
    required this.price,
    required this.quantity,
    this.size,
    this.color,
    this.image,
  });

  double get subtotal => price * quantity;

  @override
  List<Object?> get props => [lineItemId, productId, name, quantity, size, color];
}

class ShippingAddress extends Equatable {
  final String name;
  final String phone;
  final String address;
  final String city;
  final String state;
  final String pincode;

  const ShippingAddress({
    required this.name,
    required this.phone,
    required this.address,
    required this.city,
    required this.state,
    required this.pincode,
  });

  String get fullAddress => '$address, $city, $state - $pincode';

  @override
  List<Object?> get props => [name, phone, address, city, state, pincode];
}

enum OrderStatus { pending, confirmed, shipped, delivered, cancelled }
enum PaymentStatus { pending, paid, failed, refunded, partiallyRefunded }

class Order extends Equatable {
  final String id;
  final String userId;
  final List<OrderItem> items;
  final OrderStatus status;
  /// Sum of line items before coupon (falls back to [totalAmount] on older API data).
  final double subtotal;
  final double discountAmount;
  final String couponCode;
  final double totalAmount;
  final ShippingAddress shippingAddress;
  final PaymentStatus paymentStatus;
  final String? razorpayOrderId;
  final String? razorpayPaymentId;
  final DateTime createdAt;
  final DateTime updatedAt;
  final String? cancelReason;
  /// India Post / manual — admin fills when marking shipped
  final String shippingCarrier;
  final String shippingAwb;
  final String? trackingUrl;
  final DateTime? shippedAt;
  final String? razorpayRefundId;
  final String? refundStatus;
  final double? refundAmount;
  final String? refundLastError;
  final String? refundNote;

  const Order({
    required this.id,
    required this.userId,
    required this.items,
    required this.status,
    required this.subtotal,
    required this.discountAmount,
    required this.couponCode,
    required this.totalAmount,
    required this.shippingAddress,
    required this.paymentStatus,
    this.razorpayOrderId,
    this.razorpayPaymentId,
    required this.createdAt,
    required this.updatedAt,
    this.cancelReason,
    this.shippingCarrier = '',
    this.shippingAwb = '',
    this.trackingUrl,
    this.shippedAt,
    this.razorpayRefundId,
    this.refundStatus,
    this.refundAmount,
    this.refundLastError,
    this.refundNote,
  });

  bool get isPaid => paymentStatus == PaymentStatus.paid;
  bool get isCancelled => status == OrderStatus.cancelled;

  @override
  List<Object?> get props => [id];
}
